### PR TITLE
Add tag [benchmark] to node-e2e-test where performance limits are not verified

### PR DESCRIPTION
**Pull request by coufon**
This PR adds a new tag "[benchmark]" to density and resource-usage node e2e test. The performance limits will not be verified at the end of benchmark tests.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30568)

### CODE REVIEW

**Comment by coufon** [comment]
@Random-Liu  @dchen1107

File: test/e2e_node/density_test.go [hunk 1]
```diff
@@ -49,11 +49,10 @@ const (
 
 var _ = framework.KubeDescribe("Density [Serial] [Slow]", func() {
 	const (
-		// the data collection time of `resource collector' and the standalone cadvisor
-		// is not synchronizated. Therefore `resource collector' may miss data or
+		// The data collection time of resource collector and the standalone cadvisor
+		// is not synchronizated, so resource collector' may miss data or
```
Comment (Random-Liu) [inline test/e2e_node/density_test.go]: remove `'`

File: test/e2e_node/density_test.go [hunk 2]
```diff
@@ -101,98 +100,61 @@ var _ = framework.KubeDescribe("Density [Serial] [Slow]", func() {
...
+				By("Verifying latency")
+				printAndVerifyLatency(batchLag, e2eLags, itArg, true)
+
+				By("Verifying resource")
+				printAndVerifyResource(f, rc, itArg.cpuLimits, itArg.memLimits, true)
+			})
+		}
+	})
+
+	Context("create a batch of pods [benchmark]", func() {
```
Comment (Random-Liu) [inline test/e2e_node/density_test.go]: Use `[Benchmark]`, our tags usually start with capitalized charactor.

File: test/e2e_node/density_test.go [hunk 3]
```diff
@@ -217,34 +179,17 @@ var _ = framework.KubeDescribe("Density [Serial] [Slow]", func() {
...
-
-				// verify latency
 				By("Verifying latency")
-				verifyLatency(batchlag, e2eLags, itArg)
+				printAndVerifyLatency(batchlag, e2eLags, itArg, true)
 
-				// verify resource
 				By("Verifying resource")
-				verifyResource(f, itArg.cpuLimits, itArg.memLimits, rc)
+				printAndVerifyResource(f, rc, itArg.cpuLimits, itArg.memLimits, true)
```
Comment (Random-Liu) [inline test/e2e_node/density_test.go]: Why no benchmark test for sequence test? It seems that it's straight forward to add it. :)

File: test/e2e_node/resource_usage_test.go [hunk 4]
```diff
@@ -84,70 +76,114 @@ var _ = framework.KubeDescribe("Resource-usage [Serial] [Slow]", func() {
...
+		logPods(f.Client)
+	}
+
+	rc.Stop()
+
+	By("Reporting overall resource usage")
+	logPods(f.Client)
+}
+
+// printAndVerifyResource verifies whether resource usage satisfies the limit.
```
Comment (Random-Liu) [inline test/e2e_node/resource_usage_test.go]: Add some more comments like: print the resource usage as perf data.

File: test/e2e_node/resource_usage_test.go [hunk 5]
```diff
@@ -84,70 +76,114 @@ var _ = framework.KubeDescribe("Resource-usage [Serial] [Slow]", func() {
...
+		// The periodic reporting period
+		reportingPeriod = 5 * time.Minute
+
+		sleepAfterCreatePods = 10 * time.Second
+	)
+
+	rc.Start()
+
+	By("Creating a batch of Pods")
+	pods := newTestPods(testArg.pods, ImageRegistry[pauseImage], "test_pod")
```
Comment (Random-Liu) [inline test/e2e_node/resource_usage_test.go]: Why not create batch?

File: test/e2e_node/resource_usage_test.go [hunk 6]
```diff
@@ -84,70 +76,114 @@ var _ = framework.KubeDescribe("Resource-usage [Serial] [Slow]", func() {
...
-func verifyResource(f *framework.Framework, cpuLimits framework.ContainersCPUSummary,
-	memLimits framework.ResourceUsagePerContainer, rc *ResourceCollector) {
+// runResourceUsageTest runs the resource usage test
+func runResourceUsageTest(f *framework.Framework, rc *ResourceCollector, testArg resourceTest) {
+	const (
+		// The monitoring time for one test
+		monitoringTime = 10 * time.Minute
+		// The periodic reporting period
+		reportingPeriod = 5 * time.Minute
+
```
Comment (Random-Liu) [inline test/e2e_node/resource_usage_test.go]: comment here

File: test/e2e_node/resource_usage_test.go [hunk 7]
```diff
@@ -84,70 +76,114 @@ var _ = framework.KubeDescribe("Resource-usage [Serial] [Slow]", func() {
...
+			itArg := testArg
 
-				rc.Stop()
+			It(fmt.Sprintf("resource tracking for %d pods per node", itArg.pods), func() {
+				// The test collects resource usage from a standalone Cadvisor pod.
+				// The Cadvsior of Kubelet has a housekeeping interval of 10s, which is too long to
+				// show the resource usage spikes. But changing its interval increases the overhead
+				// of kubelet. Hence we use a Cadvisor pod.
+				createCadvisorPod(f)
+				rc = NewResourceCollector(containerStatsPollingPeriod)
```
Comment (Random-Liu) [inline test/e2e_node/resource_usage_test.go]: Put these in before each, and create and comment there.

File: test/e2e_node/density_test.go [hunk 8]
```diff
@@ -367,43 +416,46 @@ func newInformerWatchPod(f *framework.Framework, mutex *sync.Mutex, watchTimes m
...
+	sort.Sort(framework.LatencySlice(e2eLags))
+	return batchLag, e2eLags
+}
+
+// printAndVerifyLatency verifies that whether pod creation latency satisfies the limit.
+func printAndVerifyLatency(batchLag time.Duration, e2eLags []framework.PodLatencyData, testArg densityTest, isVerify bool) {
 	framework.PrintLatencies(e2eLags, "worst client e2e total latencies")
 
-	// Zhou: do not trust `kubelet' metrics since they are not reset!
+	// ToDo(coufon): do not trust `kubelet' metrics since they are not reset!
```
Comment (Random-Liu) [inline test/e2e_node/density_test.go]: TODO

File: test/e2e_node/density_test.go [hunk 9]
```diff
@@ -286,7 +335,7 @@ func checkPodDeleted(f *framework.Framework, podName string) error {
 	return errors.New("Pod Not Deleted")
 }
 
-// get prometheus metric `pod start latency' from kubelet
+// getPodStartLatency gets prometheus metric `pod start latency' from kubelet
```
Comment (Random-Liu) [inline test/e2e_node/density_test.go]: s/`/'

File: test/e2e_node/density_test.go [hunk 10]
```diff
@@ -367,43 +416,46 @@ func newInformerWatchPod(f *framework.Framework, mutex *sync.Mutex, watchTimes m
...
-func verifyLatency(batchLag time.Duration, e2eLags []framework.PodLatencyData, testArg densityTest) {
+// createBatchPodSequential creats pods back-to-back in sequence.
+func createBatchPodSequential(f *framework.Framework, pods []*api.Pod) (time.Duration, []framework.PodLatencyData) {
+	batchStartTime := unversioned.Now()
+	e2eLags := make([]framework.PodLatencyData, 0)
+	for _, pod := range pods {
+		create := unversioned.Now()
+		f.PodClient().CreateSync(pod)
+		e2eLags = append(e2eLags,
+			framework.PodLatencyData{Name: pod.ObjectMeta.Name, Latency: unversioned.Now().Time.Sub(create.Time)})
```
Comment (Random-Liu) [inline test/e2e_node/density_test.go]: pod.Name is enough.

File: test/e2e_node/density_test.go [hunk 11]
```diff
@@ -367,43 +416,46 @@ func newInformerWatchPod(f *framework.Framework, mutex *sync.Mutex, watchTimes m
 	return controller
 }
 
-// verifyLatency verifies that whether pod creation latency satisfies the limit.
-func verifyLatency(batchLag time.Duration, e2eLags []framework.PodLatencyData, testArg densityTest) {
+// createBatchPodSequential creats pods back-to-back in sequence.
+func createBatchPodSequential(f *framework.Framework, pods []*api.Pod) (time.Duration, []framework.PodLatencyData) {
+	batchStartTime := unversioned.Now()
+	e2eLags := make([]framework.PodLatencyData, 0)
```
Comment (Random-Liu) [inline test/e2e_node/density_test.go]: nit: var e2eLags []framework.PodLatencyData is enough

File: test/e2e_node/density_test.go [hunk 12]
```diff
@@ -257,14 +202,118 @@ type densityTest struct {
...
+		watchTimes = make(map[string]unversioned.Time, 0)
+		stopCh     = make(chan struct{})
+	)
+
+	// create test pod data structure
+	pods := newTestPods(testArg.podsNr, ImageRegistry[pauseImage], podType)
+
+	// the controller watches the change of pod status
+	controller := newInformerWatchPod(f, mutex, watchTimes, podType)
+	go controller.Run(stopCh)
```
Comment (Random-Liu) [inline test/e2e_node/density_test.go]: Use defer, or else things will not be cleaned up after fatal error such as `Failf`

File: test/e2e_node/density_test.go [hunk 13]
```diff
@@ -257,14 +202,118 @@ type densityTest struct {
...
+
+	// TODO(coufon): in the test we found kubelet starts while it is busy on something, as a result 'syncLoop'
+	// does not response to pod creation immediately. Creating the first pod has a delay around 5s.
+	// The node status has already been 'ready' so `wait and check node being ready does not help here.
+	// Now wait here for a grace period to let 'syncLoop' be ready
+	time.Sleep(sleepBeforeCreatePods)
+
+	rc.Start()
```
Comment (Random-Liu) [inline test/e2e_node/density_test.go]: defer

File: test/e2e_node/density_test.go (hunk 1 above)
Comment (coufon) [inline test/e2e_node/density_test.go]: ACK

File: test/e2e_node/density_test.go (hunk 2 above)
Comment (coufon) [inline test/e2e_node/density_test.go]: ACK

File: test/e2e_node/density_test.go (hunk 3 above)
Comment (coufon) [inline test/e2e_node/density_test.go]: ACK

File: test/e2e_node/resource_usage_test.go (hunk 4 above)
Comment (coufon) [inline test/e2e_node/resource_usage_test.go]: ACK

File: test/e2e_node/resource_usage_test.go (hunk 5 above)
Comment (coufon) [inline test/e2e_node/resource_usage_test.go]: ACK

File: test/e2e_node/resource_usage_test.go (hunk 6 above)
Comment (coufon) [inline test/e2e_node/resource_usage_test.go]: ACK

File: test/e2e_node/resource_usage_test.go (hunk 7 above)
Comment (coufon) [inline test/e2e_node/resource_usage_test.go]: ACK

File: test/e2e_node/density_test.go (hunk 8 above)
Comment (coufon) [inline test/e2e_node/density_test.go]: ACK

File: test/e2e_node/density_test.go (hunk 9 above)
Comment (coufon) [inline test/e2e_node/density_test.go]: ACK

File: test/e2e_node/density_test.go (hunk 10 above)
Comment (coufon) [inline test/e2e_node/density_test.go]: ACK

File: test/e2e_node/density_test.go (hunk 12 above)
Comment (coufon) [inline test/e2e_node/density_test.go]: ACK

File: test/e2e_node/density_test.go (hunk 13 above)
Comment (coufon) [inline test/e2e_node/density_test.go]: ACK

**Comment by Random-Liu** [comment]
LGTM

**Comment by Random-Liu** [comment]
Mark P2 because this blocks docker automated performance validation test.

**Pull request merged**